client/rest/src/main/java/org/elasticsearch/client/ResponseListener.java

```diff
@@ -19,24 +19,22 @@
 
 package org.elasticsearch.client;
 
-import org.apache.http.StatusLine;
-import org.apache.http.client.methods.CloseableHttpResponse;
-import org.apache.http.message.BasicHttpResponse;
-
-import java.io.IOException;
-
 /**
- * Simple {@link CloseableHttpResponse} impl needed to easily create http responses that are closeable given that
- * org.apache.http.impl.execchain.HttpResponseProxy is not public.
+ * Listener to be provided when calling async performRequest methods provided by {@link RestClient}.
+ * Those methods that do accept a listener will return immediately, execute asynchronously, and notify
+ * the listener whenever the request yielded a response, or failed with an exception.
  */
-class CloseableBasicHttpResponse extends BasicHttpResponse implements CloseableHttpResponse {
+public interface ResponseListener {
 
-    public CloseableBasicHttpResponse(StatusLine statusline) {
-        super(statusline);
-    }
+    /**
+     * Method invoked if the request yielded a successful response
+     */
+    void onSuccess(Response response);
```

Comment:
Looks complicated to me, I also think that the difference between `FailureListener` and `ResponseListener` is already not clear enough, and mixing them up would make things worse.

FailureListener is for getting notified whenever any request attempt fails, regardless of whether the request will finally succeed on another node. This gets set at RestClient creation. This one could be made more functional by introducing an interface for it I guess. Not sure how useful it will be as it is mainly needed internally for sniffing on failure.

ResponseListener is the listener for async requests. Will be notified on completion or failure (definitive failure only). I think this one should have two methods, not sure making it functional would make sense.

Let me know if I missed something in your suggestion, it might be that I did.